frontend/ui/trapper.lua

```diff
@@ -0,0 +1,337 @@
+--[[--
+Trapper module: provides methods for simple interaction with UI, to be
+used by linear jobs between their steps.
+
+Allows code to trap UI (give progress info to UI, ask for user choice),
+or get trapped by UI (get interrupted).
+Mostly done with coroutines, but hides their usage for simplicity.
+]]
+
+
+local ConfirmBox = require("ui/widget/confirmbox")
+local InfoMessage = require("ui/widget/infomessage")
+local UIManager = require("ui/uimanager")
+local ffiutil = require("ffi/util")
+local logger = require("logger")
+local _ = require("gettext")
+
+local Trapper = {}
+
+--[[--
+Execute function and allows it to be trapped (that is, to use our other methods)
+
+Simple wrapper function for a coroutine, which is a prerequisite
+for all our methods (this simply abstracts the @{coroutine}
+business to our callers), and execute it.
+This call should be the last step in some event processing code,
```

Comment:
Note that you'll need a separating newline if you don't want the Markdown parser to collapse this all into one paragraph.